src/protocols/openid.js

```diff
@@ -0,0 +1,144 @@
+import _ from 'lodash'
+
+import {UserModelAPI, PassportModelAPI} from '../model'
+import {config} from '../config'
+
+/**
+ * OpenID Authentication Protocol
+ *
+ * OpenID is an open standard for federated authentication.
+ *
+ * For more information on OpenID in Passport.js, check out:
+ * http://www.passportjs.org/packages/passport-openidconnect/
+ *
+ */
+
+function parseJwt(token) {
+  return JSON.parse(Buffer.from(token.split('.')[1], 'base64').toString())
+}
+
+function constructUserInfo(profile, accessToken) {
+  let user = {}
+
+  const {clientId} = config.api.keycloak
+  const {sub, resource_access} = parseJwt(accessToken)
+
+  if (resource_access && resource_access[`${clientId}`]) {
+    user.groups = resource_access[`${clientId}`].roles
+  }
+
+  if (profile.emails && profile.emails[0]) {
+    user.email = profile.emails[0].value
+  } else if (profile.username) {
+    user.email = profile.username
+  }
+
+  user.firstname = _.has(profile, 'name.givenName')
+    ? profile.name.givenName
+    : profile.username
+  user.surname = _.has(profile, 'name.familyName')
+    ? profile.name.familyName
+    : profile.username
+
+  user.provider = 'keycloak'
+
+  // In case the user was created before
+  user.token = null
+  user.tokenType = null
+  user.locked = false
+  user.expiry = null
+
+  return {user, identifier: sub}
+}
+
+function createOrUpdateUser(user) {
+  return UserModelAPI.findOne({email: user.email}).then(function (_user) {
```

Comment:
Minor - rename the parameter for the function to something else. The underscore is used to indicate that a parameter is not used in the function.

Reply:
cool done :)